### ISSUE TITLE
Missing type declarations

### ISSUE THREAD

**Issue by JstnMcBrd** [jsx-eslint/eslint-plugin-react#3776]
### Is there an existing issue for this?

- [X] I have searched the existing issues and my issue is unique
- [X] My issue appears in the command-line and not only in the text editor

### Description Overview

When imported in a TypeScript environment, `eslint-plugin-react` throws a "missing type declarations" error.

![image](https://github.com/jsx-eslint/eslint-plugin-react/assets/28303477/7812e310-ff0d-43f0-b73f-0157e8188728)

### Why

This is because `eslint-plugin-react` does not have any type declarations included in the package. These would usually be in an `index.d.ts` file.

### Expected Behavior

The expected behavior is no errors. Type declarations would also be convenient for intellisense and type hints - like easily communicating to users what the available config types are, for example.

![image](https://github.com/jsx-eslint/eslint-plugin-react/assets/28303477/337e477a-5964-4974-a27b-e32d388e38a7)

### Workaround

Users can follow the instructions in the error message and make a temporary module augmentation to declare their own types for the package.

In my personal project, I've done it this way:

```ts
// ./src/@types/eslint-plugin-react.d.ts
declare module 'eslint-plugin-react' {
	import type { ESLint, Linter } from 'eslint';
	const plugin: Omit<ESLint.Plugin, 'configs'> & {
		// eslint-plugin-react does not use FlatConfig yet
		configs: Record<string, Linter.Config>;
	};
	export default plugin;
}
```

### Solution

Personally, I'd always recommended writing everything in TypeScript, and then type declaration files will be automatically generated for you :)

But the quick and easy solution would be to write your own `index.d.ts` declaration file in the project root.

I would be happy submit a PR for this myself, if it helps.

### eslint-plugin-react version

v7.34.3 (but technically all versions)

### eslint version

v8.57.0 (irrelevant)

### node version

v20.13.1 (irrelevant)

**Comment by ljharb** [comment]
It's simply not a reasonable expectation that an npm package ships TS types, nor necessarily has a DT package. "Write everything in TS" isn't a universally palatable choice.

At some point in the future, we may ship d.ts files, but in the meantime, the proper solution (for every npm package that doesn't ship types) is to make a DT package.
